src/backend/num_bigint.rs

```diff
@@ -136,14 +140,21 @@ impl Integer {
         self.pow_mod_ref(exponent, modulo)
     }
     pub fn pow_mod_ref(&self, exponent: &Self, modulo: &Self) -> Option<Self> {
-        if exponent.0.is_negative() {
+        let r = if exponent.0.is_negative() {
             let nself = self.0.modinv(&modulo.0)?;
             Some(Integer(nself.modpow(&-&exponent.0, &modulo.0)))
         } else {
             Some(Integer(self.0.modpow(&exponent.0, &modulo.0)))
+        };
+        if modulo.cmp0().is_lt() && r.as_ref().map(|r| r.cmp0().is_ne()).unwrap_or(false) {
+            eprintln!("adjusting {r:?}");
+            r.map(|r| r - modulo)
+        } else {
+            r
         }
```

Comment:
maybe you'll find this way of writing it more readable?
```suggestion
        match r {
            Some(r) if r.cmp0().is_ne() && modulo.cmp0().is_lt() => {
                Some(r - modulo)
            }
            Some(r) => Some(r),
            None => None,
        }
```

Reply:
or last two branches can be replaced with `x => x`

Reply:
Actually I can just remove all of maps by using `?` one line above, and this becomes a lot more readable